lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.UUID;
+
+import com.palantir.lock.v2.LockToken;
+
+class SharedImmutableTsToken implements LockToken {
+    private LockToken token;
+    private volatile int reference;
+
+    SharedImmutableTsToken(LockToken token) {
```

Comment:
think it's worth this taking the reference count in its constructor and only allowing it to be decremented. It means you can't get to zero and then go back up to 1